### ISSUE TITLE
Incorrect `unsupported-operator` with operator `in` on types `str` and `T & str`

### ISSUE THREAD

**Issue by shilch** [astral-sh/ty#485]
### Summary

Minimal sample:

```python
def do_something[T](val: T, s: str):
    if type(val) is str:
        if s in val:
            print('s in val')
```

Error by ty:

```
error[unsupported-operator]: Operator `in` is not supported for types `str` and `T`, in comparing `str` with `T & str`
 --> test.py:3:12
  |
1 | def do_something[T](val: T, s: str):
2 |     if type(val) is str:
3 |         if s in val:
  |            ^^^^^^^^
4 |             print('s in val')
  |
info: rule `unsupported-operator` is enabled by default
```

### Version

ty 0.0.1-alpha.6

**Comment by sharkdp** [comment]
Thank you for reporting this.

It seems like we mistakenly emit an `unsupported-operator` diagnostic if *any* of the members in a the intersection do not support the operator. But we should only report that if *none* of the members support it, I think.

More self-contained example:
```py
class Container:
    def __contains__(self, x) -> bool:
        return True

class NonContainer:
    pass

def f(x: Container):
    if isinstance(x, NonContainer):
        reveal_type(x)  # Container & NonContainer

        "a" in x  # should not be an error
```
https://play.ty.dev/8d1dabca-2623-4431-b08c-963d9e48ebb2

**Comment by sharkdp** [comment]
This needs to be fixed in `TypeInferenceBuilder::infer_binary_intersection_type_comparison` where we currently short-circuit on every error.


We even have a test for this here, which asserts that there should be an error, but I think that is wrong: https://github.com/astral-sh/ruff/blob/6df10c638e3afed4a3fd9145d0353861e29d6acc/crates/ty_python_semantic/resources/mdtest/comparison/intersections.md?plain=1#L127

Ideally, we would want to distinguish between "operator not available" and "call to operator failed". The former should not be an error (if other intersection elements lead to a successful call), but the latter should probably always be an error?
